docs: Update local-test-setup.md to use scripts (#2178)
kyma-project/lifecycle-manager: docs/contributor/04-local-test-setup.md

> # Local Test Setup in the Control Plane Mode Using k3d

Author: As of now we only support k3d, so I think we can omit it as you suggested.

* Strict mTLS connection between Kyma Control Plane (KCP) and SKR clusters
* SAN Pinning (SAN of client TLS certificate needs to match the DNS annotation of a corresponding Kyma CR)

## Prerequisites

The following tooling is required in the versions defined in [`versions.yaml`](../../versions.yaml):

- docker
- go
- golangci-lint
- istioctl
- k3d
- kubectl
- kustomize
- [modulectl](https://github.com/kyma-project/modulectl)
- yq

## Procedure

Execute the following scripts from the project root.

### 1. Create Test Clusters

Create local test clusters for SKR and KCP.

```sh
K8S_VERSION=$(yq e '.k8s' ./versions.yaml)
CERT_MANAGER_VERSION=$(yq e '.certManager' ./versions.yaml)
./scripts/tests/create_test_clusters.sh --k8s-version $K8S_VERSION --cert-manager-version $CERT_MANAGER_VERSION
```

### 2. Install the CRDs

> Install the CRDs to the KCP cluster.

Author: Let's link to the other doc

```sh
./scripts/tests/install_crds.sh
```

### 3. Deploy lifecycle-manager

> #### 3.1 Deploy lifecycle-manager from a Registry

Reviewer: ```suggestion
You can deploy Lifecycle Manager either from the registry or local sources. Choose one of the below options:

3.1 Deploy a built image from the registry, for example, the `latest` image from the `prod` registry.
```
It could look cleaner if we don't use H4 but just an ordered list. Then the title case is not necessary and we can write the steps straight away.

> Install the [Lifecycle Manager CRDs](./resources/README.md) to the KCP cluster.

Reviewer: ```suggestion
Install the [Lifecycle Manager CRDs](./resources/README.md) in the KCP cluster.
```

> Deploy a built image from the registry, e.g. the `latest` image from the `prod` registry.

Reviewer: ```suggestion

```
Moved to the line above.

```sh
REGISTRY=prod
TAG=latest
./scripts/tests/deploy_klm_from_registry.sh --image-registry $REGISTRY --image-tag $TAG

> ```

Reviewer: Please indent the command.

> #### 3.2 Deploy lifecycle-manager from Local Sources

Reviewer: ```suggestion
3.2 Build a new image from the local sources, push it to the local KCP registry, and deploy it.
```

> Build a new image from the local sources, push it to the local KCP registry and deploy it.

Reviewer: ```suggestion

```
Moved to the line above.

```sh
./scripts/tests/deploy_klm_from_sources.sh

> ```

Reviewer: Please indent the command.

### 4. Deploy a Kyma CR

```sh
SKR_HOST=host.k3d.internal
./scripts/tests/deploy_kyma.sh $SKR_HOST
```

> ### 5. Verify If The Kyma Becomes Ready

Reviewer: ```suggestion
### 5. Verify If the Kyma CR Becomes Ready
```

> #### 5.1 Verify If Kyma Is Ready in KCP (Takes Roughly 1–2 Minutes)

Reviewer: ```suggestion
5.1 Verify if the Kyma CR is in the `Ready` state in KCP. It takes roughly 1–2 minutes.
```

```sh
kubectl config use-context k3d-kcp
kubectl get kyma/kyma-sample -n kcp-system

> ```

Reviewer: Please indent the command.

> #### 5.1 Verify If Kyma Is Ready in SKR (Takes Roughly 1-2 Minutes)

Reviewer: ```suggestion
5.1 Verify if the Kyma CR is in the `Ready` state in SKR. It takes roughly 1-2 minutes.
```

```sh
kubectl config use-context k3d-skr
kubectl get kyma/default -n kyma-system

> ```

Reviewer: Please indent the command.

### 6. [OPTIONAL] Deploy template-operator Module

> Build the template-operator module from the local sources, push it to the local KCP registry and deploy it.

Reviewer: ```suggestion
Build the template-operator module from the local sources, push it to the local KCP registry, and deploy it.
```

> ### 6. [OPTIONAL] Deploy template-operator Module

Reviewer: ```suggestion
### 6. [OPTIONAL] Deploy the template-operator Module
```

```sh
cd <template-operator-repository>

make build-manifests
modulectl create --config-file ./module-config.yaml --registry http://localhost:5111 --insecure 

kubectl config use-context k3d-kcp
# repository URL is localhost:5111 on the host machine but must be k3d-kcp-registry.localhost:5000 within the cluster
yq e '.spec.descriptor.component.repositoryContexts[0].baseUrl = "k3d-kcp-registry.localhost:5000"' ./template.yaml | kubectl apply -f -

MT_VERSION=$(yq e '.spec.version' ./template.yaml)
cd <lifecycle-manager-repository>
./scripts/tests/deploy_modulereleasemeta.sh template-operator regular:$MT_VERSION
```

### 7. [OPTIONAL] Add the template-operator Module to the Kyma CR and Verify If It Becomes Ready

> #### 7.1 Add the Module to the Kyma CR Spec

Reviewer: ```suggestion
7.1 Add the template-operator module to the Kyma CR spec.
```

```sh
kubectl config use-context k3d-skr
kubectl get kyma/default -n kyma-system -o yaml | yq e '.spec.modules[0]={"name": "template-operator"}' | kubectl apply -f -

> ```

Reviewer: Please indent the command.

> #### 7.2 Verify If the Module Becomes Ready (Takes Roughly 1–2 Minutes)

Reviewer: ```suggestion
7.2 Verify if the module becomes `Ready`. It takes roughly 1–2 minutes.
```

```sh
kubectl config use-context k3d-skr
kubectl get kyma/default -n kyma-system -o wide

> ```

Reviewer: Please indent the command.

> #### 7.3 Remove the Module from the Kyma CR Spec

Reviewer: ```suggestion
7.3 Remove the module from the Kyma CR spec.
```

```sh
kubectl config use-context k3d-skr
kubectl get kyma/default -n kyma-system -o yaml | yq e 'del(.spec.modules[0])' | kubectl apply -f -

> ```

Reviewer: Please indent the command.

### 8. [OPTIONAL] Verify Conditions

> Check the conditions of the Kyma.

Reviewer: Kyma cluster? Kyma CR?

> Check the conditions of the Kyma.

Author: ```suggestion
Check the conditions of the Kyma CR in the KCP cluster.
```

- `SKRWebhook` to determine if the webhook has been installed to the SKR

> - `ModuleCatalog` to determine if the ModuleTemplates and ModuleReleaseMetas haven been synced to the SKR cluster

Reviewer: ```suggestion
- `ModuleCatalog` to determine if the ModuleTemplate CRs and ModuleReleaseMeta CRs haven been synced to the SKR cluster
```

> - `Modules` to determine if the added modules are ready

Reviewer: ```suggestion
- `Modules` to determine if the added modules are `Ready`
```
do you find it correct?

```sh
kubectl config use-context k3d-kcp
kubectl get kyma/kyma-sample -n kcp-system -o yaml | yq e '.status.conditions'
```

### 9. [OPTIONAL] Verify If Watcher Events Reach KCP

> #### 9.1 Flick the Channel to Trigger an Event

Reviewer: 9.1 Flick the channel to trigger an event.
```

```sh
kubectl config use-context k3d-skr
kubectl get kyma/default -n kyma-system -o yaml | yq e '.spec.channel="regular"' | kubectl apply -f -
kubectl get kyma/default -n kyma-system -o yaml | yq e '.spec.channel="fast"' | kubectl apply -f -

> ```

Reviewer: Please indent the command.

> #### 9.2 Verify if lifecyle-manger Received the Event on KCP

Reviewer: ```suggestion
9.2 Verify if Lifecycle Manger received the event in KCP.
```

```sh
kubectl config use-context k3d-kcp
kubectl logs deploy/klm-controller-manager -n kcp-system | grep "event received from SKR"

> ```

Reviewer: Please indent the command.

> #### 10. [OPTIONAL] Delete the Local Test Clusters

Reviewer: ```suggestion
### 10. [OPTIONAL] Delete the Local Test Clusters
```

* Strict mTLS connection between Kyma Control Plane (KCP) and SKR clusters
* SAN Pinning (SAN of client TLS certificate needs to match the DNS annotation of a corresponding Kyma CR)

## Prerequisites

The following tooling is required in the versions defined in [`versions.yaml`](../../versions.yaml):

> - cmctl (cert-manager)

Author: Yes right, we can remove that one from the list 👍🏻

> - cmctl (cert-manager)

Author: ```suggestion
```

- docker
- go
- golangci-lint
- istioctl
- k3d
- kubectl
- kustomize
- [modulectl](https://github.com/kyma-project/modulectl)
- yq

## Procedure

Execute the following scripts from the project root.

## Create Test Clusters

Create local test clusters for SKR and KCP.

```sh
K8S_VERSION=$(yq e '.k8s' ./versions.yaml)
CERT_MANAGER_VERSION=$(yq e '.certManager' ./versions.yaml)
./scripts/tests/create_test_clusters.sh --k8s-version $K8S_VERSION --cert-manager-version $CERT_MANAGER_VERSION
```

## Install the CRDs

Install the CRDs to the KCP cluster.

```sh
./scripts/tests/install_crds.sh
```

## Deploy lifecycle-manager

Deploy a built image from the registry, e.g. the `latest` image from the `prod` registry.

```sh
REGISTRY=prod
TAG=latest
./scripts/tests/deploy_klm_from_registry.sh --image-registry $REGISTRY --image-tag $TAG
```

OR build a new image from the local sources, push it to the local KCP registry and deploy it.

```sh
./scripts/tests/deploy_klm_from_sources.sh
```

## Deploy a Kyma CR

```sh

> SKR_HOST=host.k3d.internal

Reviewer 2: I think we need to handle the option of using the local Lifecycle Manager, where the secret should just be:
`config: $(k3d kubeconfig get skr | base64 | tr -d '\n')`

Author: Yes, for that `SKR_HOST` would need to be set to `localhost`. I didn't explain it explicitly as it is described as part of the script we call.
